asynciterator.ts

```diff
@@ -777,9 +789,95 @@ export class IntegerIterator extends AsyncIterator<number> {
   }
 }
 
+/**
+ * A synchronous mapping function from one element to another.
+ * The iterator performing the mapping is passed as a second argument.
+ */
+export type MapFunction<S, D = S> = (item: S) => D | null;
+
+/**
+ An iterator that calls a synchronous mapping function
+ on every item from its source iterator.
+ @extends module:asynciterator.AsyncIterator
+*/
+export class MappingIterator<S, D = S> extends AsyncIterator<D> {
+  private readonly _destroySource: boolean;
+  protected _source: InternalSource<S>;
+
+  /**
+   * Applies the given mapping to the source iterator.
+   */
+  constructor(
+    _source: AsyncIterator<S>,
+    private _map: MapFunction<S, D>,
+    options: SourcedIteratorOptions = {}
+  ) {
+    super();
+    this._source = _validateSource(_source);
+    this._destroySource = options.destroySource !== false;
+
+    if (_source.done) {
+      this.close();
+    }
+    else {
+      this._source._destination = this;
+      this._source.on('end', destinationClose);
+      this._source.on('error', destinationEmitError);
+      this._source.on('readable', destinationSetReadable);
+      // If we are given a source that is already readable
+      // then we need to set the state of this iterable to readable
+      // also as there is no guarantee that the is no forthcoming
+      // readable event from the source
+      this.readable = this._source.readable;
+    }
+  }
+
+  read(): D | null {
+    // Do not read the source if the current iterator is ended
+    if (this.done)
+      return null;
+
+    // Do not read from the source if it is ended
+    // TODO: Possibly remove this after https://github.com/RubenVerborgh/AsyncIterator/pull/45 is merged
+    // the requirement for this to be removed is that the 'end' emit occurs synchronously when the ENDED
+    // state is set, which does not currently occur if endAsync is called.
+    if (this._source.done) {
+      this.close();
+      return null;
+    }
```

Comment:
```suggestion
    if (!this._source.readable) {
      this.readable = false;
      if (this._source.done)
        this.close();
      return null;
    }
```

^ We should be maximally well-behaved and not attempt to call `.read` on the source unless it advertises itself as potentially readable.

Reply:
@RubenVerborgh Is it worth opening an issue to apply this idea to other iterators, for instance in [`_readAndTransform`](https://github.com/RubenVerborgh/AsyncIterator/blob/a536bca2dfe84697e1a095c10eb1fcce40f7182c/asynciterator.ts#L1185) and [`_readAndTransformSimple`](https://github.com/RubenVerborgh/AsyncIterator/blob/a536bca2dfe84697e1a095c10eb1fcce40f7182c/asynciterator.ts#L1330-L1339)

Reply:
Note that `TransformIterator` and `SimpleTransformIterator` also accept regular Node.js `Readable` instances, which do not define `readable`. So the goal there is to also work with non-`AsyncIterator`s (see `wrap`).

We can have the same goal for `MappingIterator`, but that requires different tests.